File: SecurePassEngine/Sources/SecurePassEngine/SecurePassEngine.swift
@_silgen_name("log")
private func libcLog(_ value: Double) -> Double

public enum SecurePassError: Int32 {
    case success = 0
    case invalidParameter = -1
    case bufferTooSmall = -2
}

public enum SecurePassEngine {

    private static let lowercase = Array("abcdefghijklmnopqrstuvwxyz")
    private static let uppercase = Array("ABCDEFGHIJKLMNOPQRSTUVWXYZ")
    private static let digits = Array("0123456789")
    private static let special = Array("!@#$%^&*()-_=+[]{}|;:,.<>?")

    public static func generatePassword(length: Int, includeSpecialChars: Bool) -> String? {
        guard length > 0 else { return nil }

        var pools: [[Character]] = [lowercase, uppercase, digits]
        if includeSpecialChars {
            pools.append(special)
        }

        let charset = pools.flatMap { $0 }
        guard !charset.isEmpty else { return nil }

        var rng = SystemRandomNumberGenerator()
        var password = [Character](repeating: charset[0], count: length)

        let guaranteedCount = min(length, pools.count)
        for index in 0..<guaranteedCount {
            password[index] = pools[index].randomElement(using: &rng)!
        }

        for index in guaranteedCount..<length {
            password[index] = charset.randomElement(using: &rng)!
        }

        for index in stride(from: length - 1, through: 1, by: -1) {
            let swapIndex = Int.random(in: 0...index, using: &rng)
            password.swapAt(index, swapIndex)
        }

        return String(password)
    }

    public static func calculateEntropy(password: String) -> Double {
        guard !password.isEmpty else { return 0 }

        var poolSize = 0
        var hasLower = false
        var hasUpper = false
        var hasDigit = false
        var hasSpecial = false

        for scalar in password.unicodeScalars {
            switch scalar.value {
            case 97...122: hasLower = true
            case 65...90: hasUpper = true
            case 48...57: hasDigit = true
            default: hasSpecial = true
            }
        }

        if hasLower { poolSize += lowercase.count }
        if hasUpper { poolSize += uppercase.count }
        if hasDigit { poolSize += digits.count }
        if hasSpecial { poolSize += special.count }

        guard poolSize > 0 else { return 0 }

        return Double(password.count) * entropyLog2(Double(poolSize))
    }

    public static func charsetSize(includeSpecialChars: Bool) -> Int {
        var size = lowercase.count + uppercase.count + digits.count
        if includeSpecialChars {
            size += special.count
        }
        return size
    }

    private static func entropyLog2(_ value: Double) -> Double {
        guard value > 0 else { return 0 }
        return libcLog(value) / libcLog(2.0)
    }
}

@_cdecl("securepass_generate_password")
public func securepass_generate_password(
    _ length: Int32,
    _ includeSpecial: Int32,
    _ buffer: UnsafeMutablePointer<CChar>?,
    _ bufferSize: Int32
) -> Int32 {
    guard let buffer, bufferSize > 0, length > 0 else {
        return SecurePassError.invalidParameter.rawValue
    }

    let includeSpecialChars = includeSpecial != 0
    guard let password = SecurePassEngine.generatePassword(
        length: Int(length),
        includeSpecialChars: includeSpecialChars
    ) else {
        return SecurePassError.invalidParameter.rawValue
    }

    let utf8 = Array(password.utf8CString)
    guard utf8.count <= bufferSize else {
        return SecurePassError.bufferTooSmall.rawValue
    }

    for index in 0..<utf8.count {
        buffer[index] = utf8[index]
    }

    return SecurePassError.success.rawValue
}

@_cdecl("securepass_calculate_entropy")
public func securepass_calculate_entropy(_ password: UnsafePointer<CChar>?) -> Double {
    guard let password else { return 0 }
    return SecurePassEngine.calculateEntropy(password: String(cString: password))
}

@_cdecl("securepass_charset_size")
public func securepass_charset_size(_ includeSpecial: Int32) -> Int32 {
    Int32(SecurePassEngine.charsetSize(includeSpecialChars: includeSpecial != 0))
}

File: SecurePassAndroid/app/src/main/cpp/securepass_engine.c
#include "SecurePassEngine.h"

#include <math.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>

static const char LOWERCASE[] = "abcdefghijklmnopqrstuvwxyz";
static const char UPPERCASE[] = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
static const char DIGITS[] = "0123456789";
static const char SPECIAL[] = "!@#$%^&*()-_=+[]{}|;:,.<>?";

static uint32_t random_below(uint32_t upper) {
    if (upper == 0) {
        return 0;
    }

    uint32_t value = 0;
    arc4random_buf(&value, sizeof(value));
    return value % upper;
}

static void shuffle_chars(char *buffer, int32_t length) {
    for (int32_t index = length - 1; index > 0; --index) {
        int32_t swap_index = (int32_t) random_below((uint32_t) (index + 1));
        char temp = buffer[index];
        buffer[index] = buffer[swap_index];
        buffer[swap_index] = temp;
    }
}

static char pick_from_pool(const char *pool) {
    return pool[random_below((uint32_t) strlen(pool))];
}

int32_t securepass_generate_password(
        int32_t length,
        int32_t include_special,
        char *buffer,
        int32_t buffer_size) {
    if (buffer == NULL || buffer_size <= 0 || length <= 0) {
        return -1;
    }
    if (length + 1 > buffer_size) {
        return -2;
    }

    const char *pools[4];
    int pool_count = 3;
    pools[0] = LOWERCASE;
    pools[1] = UPPERCASE;
    pools[2] = DIGITS;
    if (include_special != 0) {
        pools[3] = SPECIAL;
        pool_count = 4;
    }

    char charset[128];
    size_t charset_len = 0;
    for (int i = 0; i < pool_count; ++i) {
        size_t pool_len = strlen(pools[i]);
        if (charset_len + pool_len >= sizeof(charset)) {
            return -1;
        }
        memcpy(charset + charset_len, pools[i], pool_len);
        charset_len += pool_len;
    }

    int32_t guaranteed_count = length < pool_count ? length : pool_count;
    for (int32_t index = 0; index < guaranteed_count; ++index) {
        buffer[index] = pick_from_pool(pools[index]);
    }
    for (int32_t index = guaranteed_count; index < length; ++index) {
        buffer[index] = charset[random_below((uint32_t) charset_len)];
    }

    shuffle_chars(buffer, length);
    buffer[length] = '\0';
    return 0;
}

int32_t securepass_charset_size(int32_t include_special) {
    int32_t size = (int32_t) (sizeof(LOWERCASE) + sizeof(UPPERCASE) + sizeof(DIGITS) - 3);
    if (include_special != 0) {
        size += (int32_t) (sizeof(SPECIAL) - 1);
    }
    return size;
}

double securepass_calculate_entropy(const char *password) {
    if (password == NULL || password[0] == '\0') {
        return 0.0;
    }

    int has_lower = 0;
    int has_upper = 0;
    int has_digit = 0;
    int has_special = 0;

    for (const unsigned char *cursor = (const unsigned char *) password; *cursor != '\0'; ++cursor) {
        if (*cursor >= 'a' && *cursor <= 'z') {
            has_lower = 1;
        } else if (*cursor >= 'A' && *cursor <= 'Z') {
            has_upper = 1;
        } else if (*cursor >= '0' && *cursor <= '9') {
            has_digit = 1;
        } else {
            has_special = 1;
        }
    }

    int pool_size = 0;
    if (has_lower) {
        pool_size += (int) (sizeof(LOWERCASE) - 1);
    }
    if (has_upper) {
        pool_size += (int) (sizeof(UPPERCASE) - 1);
    }
    if (has_digit) {
        pool_size += (int) (sizeof(DIGITS) - 1);
    }
    if (has_special) {
        pool_size += (int) (sizeof(SPECIAL) - 1);
    }

    if (pool_size == 0) {
        return 0.0;
    }

    return (double) strlen(password) * log2((double) pool_size);
}
